### ISSUE TITLE
data: dates in image DataTable aren't supported by SQLite

### ISSUE THREAD

**Issue by twest820** [CascadesCarnivoreProject/Carnassial#81]
Date column's format is dd-MMM-yyyy (Constants.Time.DateFormat) but SQLite's date functions require yyyy-MM-dd.  
- select \* from DataTable where time(Time) >= time("12:00:00") works
- select \* from DataTable where date(Date) >= date("2016-06-01") returns no rows though many exist
- select \* from DataTable where strftime("dd-MMM-yyyy HH:mm:ss", Date || " " || Time) >= datetime("2016-06-01 00:00:00") also returns no rows

Preferred fix is to replace the Date and Time columns with something like

```
ALTER TABLE DataTable ADD COLUMN Timestamp DATETIME
```

Increasing the precision from seconds to milliseconds as part of this change is also desirable as current generation cameras are capable of multiple still images per second.  See http://www.sqlite.org/lang_datefunc.html.

A related difficulty is SQLite does not support time zone indicators, meaning either a separate timezone column is required or UTC times must be used.  This is needed to handle cases where cameras aren't in the same timezone as Timelapse is running in due to either physical distance or cases such as cameras following solar noon (standard time) in timezones where daylight savings time is used.

Blocks most performant completion of #80 as rows have to be filtered in code rather than via database query.

**Comment by saulgreenberg** [comment]
[Copy of email sent]
Ah, this requires discussion. 
Up front, I think the idea of storing date/time in a proper SQL format is a good one, but it needs to be done very very carefully.

The reason I implement the dd-MMM-yyyy format is that it removes all ambiguities about what the day vs month is, i.e. There is no ambiguity in 02-Jan-2016, but there is in 2016-01-02. Other date formats I have seen are even worse: eg., 11-02-12 . I have gone through lengths to try to figure out the date format when reading in images, in notifying the user when the software could not remove ambiguities, and even in allowing users to correct the most common ambiguity issues (e.g., swapping month/day).

Even after the dates are read in, the way they are presented to the user (both in the interface and in any output) remains very important. For example, the yyyy-MM-dd format is inherently ambiguous  if the user does not know that date format ahead of time i.e., if the date is written as 2016-01-02,  they don’t know if its January 2 or February 1st. 

Remember too that this system is used internationally, on computers where users have set a variety of date formats  (via the Windows control panel) which can introduce nasty problems interpreting dates if there is any potential for date ambiguity (I worked with various users to identify and solve those problems).

Of course, if we centralize how dates are interpreted in one spot, we can translate dates into whatever format we want i.e. if a date is translated into yyyy-MM-dd when storing it into the database, but then translated back to dd-MMM-yyyy format when read back out (e.g., to present it in the interface). However, this will still leave in some ambiguity to users who choose to read the database directly for analysis (and some of them do) – While I suppose a knowledgeable user would understand how SQL stores dates, I am sure there are some who will screw up: at the very least, we would need to document that in the manual.

Other issues that would be introduced in date changes are:
-   Backwards compatibility (for users who use the existing format)
-   Enforcement of a date format in the date field (i.e., so arbitrary text or dates not adhering to that format cannot be entered).
-   The annoying issue of Excel changing the date format when exporting (and perhaps re-introducing ambiguity)

A bit of history: the date dialogs I have I the software are those that users have requested, and no one has asked for any others. While they are not generalized date-changing functions, their language and action mirrors exact requests that users have asked for – none have asked for a generalized data-changing function. 

Another thing to consider is interface complexity for some of your changes, i.e., specifying date ranges for date corrections through the UI adds complexity. A different approach (and one which would allow us to deal with dates without having to query on dates) is to have the user select the beginning and end image for a date operation (or perhaps the image folder in your case), where that operation is applied to the dates for all images between those two. 

As for daylight saving times, some cameras include corrections within it, others don’t. The dialog I include for that allows users to adjust that, and they appear to be quite happy with it. However, no one has reported camera time drift (which is why I don’t deal with that), and of course we didn’t have multiple cameras (as could now be done using subfolders)

I arrived back home this evening. I should be around the next few days if you want to chat. This is definitely worth discussing before doing anything on it... and I would suggest waiting until after the next release as there are large ramifications to this.
